Refactored SBS + BQ40Z80 battery drivers

amikhalev opened the pull request:
**Describe problem solved by this pull request**
`batt_smbus` is difficult to extend for different hardware.

**Describe your solution**
Write a generic base class `SBSBattery` for SBS-compatible battery drivers. 
Implement a generic SBS as well as BQ40Z80 driver based on this.
Remove `batt_smbus` and replace with this.

**Test data / coverage**
Tested with a BQ40Z80 battery. We at least know basic information reporting works

![image](https://user-images.githubusercontent.com/4969631/119733062-3dd7db00-be36-11eb-9643-8c74147e32fd.png)

mhkabir commented:
@dagar We're also planning to separate out the largely static data (e.g manufacturing date, state of health estimate, etc.) into a separate message called `battery_info` which is published at some very low rate (e.g 0.5hz)

I'll update this post with a proposed split

mhkabir commented:
@echoGee @nbelanger99 This should let you add-on functionality for Batmon very easily, with very few lines of additional code. See the BQ40Z80 driver which extends the SBS base driver for reference.

We also tested the `sbs_generic` driver standalone, and it works out of the box with any battery which supports the SBS standard

dagar commented:
Can you rebase and we'll get this in?

amikhalev commented:
I can rebase in the next few days

amikhalev committed:
Add refactored SBS + BQ40Z80 battery drivers

 * Revised SBS base class `SBSBattery` for improved abstraction.
 * Add generic driver for generic SBS smart batteries using
 `SBSBattery`.
 * Add driver for BQ40Z80 batteries using `SBSBattery`.

Signed-off-by: Alex Mikhalev <alex@corvus-robotics.com>

amikhalev committed:
smbus: Improve error handling and comments

Signed-off-by: Alex Mikhalev <alex@corvus-robotics.com>

julianoes committed:
batmon: Adapt after smbus_sbs changes

With the refactor of smbus_sbs, most functionality has been moved to the
base class. Remaining in this driver is getting the cell count, reading
out the individual cells, and calculating the max delta between cells.

This commit removes suspend and resume functionality for now assuming
that's not actually used.

Signed-off-by: Alex Mikhalev <alex@corvus-robotics.com>

julianoes committed:
Remove batt_smbus

This has been replaced by the sbs_generic, as well as the derived
batmon and bq40z80 drivers.

Signed-off-by: Alex Mikhalev <alex@corvus-robotics.com>

julianoes committed:
smart_battery: add common smart battery Kconfig

Signed-off-by: Alex Mikhalev <alex@corvus-robotics.com>

julianoes committed:
boards: remove batt_smbus, build all new ones

This removes the removed BATT_SMBUS and replaces it with the groupd of
SMART_BATTERIES which contains sbs_generic, batmon, and bq40z80.

Signed-off-by: Alex Mikhalev <alex@corvus-robotics.com>

amikhalev commented:
This PR has been rebased on master and updated with some more changes. It is ready for review again @dagar.

Since it also involves changes to the batmon driver, could @echoGee take a look as well?

dagar commented:
There's quite a lot to reconcile here, I also have some changes in this area. I'll start digging into this post v1.13 (probably next week).

dagar commented:
The SBS generic configurable cell count (SBS_BAT_N_CELLS) feels kind of wrong. I think I'd rather change/fix our battery_status usage so we aren't required to fake something we don't necessarily need.
